src/artofillusion/polymesh/QuadMesh.java

```diff
@@ -1178,8 +1178,18 @@ private void smoothMesh(double tol, boolean calcProjectedEdges, int ns, int maxN
                     ++hardnum;
                 }
             }
-            pos.scale(1.0 / ((double) count * count));
-            pos.add(vertices[i].r.times(1.0 - 3.0 / (2.0 * count) - 1.0 / (4.0 * count)));
+
+            // In some rare cases 'count' may be 0 and 'pos' may be Vec3[0.0, 0.0, 0.0].
+            // This would result Vec3 values [Nan, Nan, Nan] or [Infinity, Infinity, Infinity]
+
+            if (count != 0)
+                pos.scale(1.0 / ((double) count * count));
+
+            if (count == 0)
+                pos.add(vertices[i].r);
+            else
+                pos.add(vertices[i].r.times(1.0 - 3.0 / (2.0 * count) - 1.0 / (4.0 * count)));
```

Comment:
I believe that this would be more simply expressed (both in terms of code run complexity and readability) as:

````
            if (count == 0) //No hard edges, don't scale.
                pos.add(vertices[i].r);
            else
            {
                pos.scale(1.0 / ((double) count * count));
                pos.add(vertices[i].r.times(1.0 - 3.0 / (2.0 * count) - 1.0 / (4.0 * count)));
            }
````
Note that this doesn't actually guard against the `pos = [0.0, 0.0, 0.0]` case - either as you've written it, or as I'm suggesting. Might still get infinities. What would the correct action be if `pos` is a Zero Vec?

(It would really be lovely if we had some documentation on what some of these variables actually mean. Do you have references? I don't recall Francois talking about it in the polymesh manual.)

Reply:
I'll need to have another look, it's been a while but it must have made sense to me at the time.

Anyway `pos = [0.0, 0.0, 0.0]` is still just a point in space, you can scale it by any real number. It just does not change.

Reply:
The history with the structure is, that I just added a pre-check of their own to both of the lines on the old code. In a way there are two separate things happening. I probably thought of combining the logic, but the decided that the change was clearer this way.